include/mfmg/dealii/dealii_matrix_free_mesh_evaluator.hpp

```diff
@@ -17,36 +17,106 @@
 
-  dealii::LinearAlgebra::distributed::Vector<double>
-  get_diagonal_inverse() const;
+  // FIXME get rid of template argument and make it virtual so that it
+  // eventually becomes the customization point that the user overrides
+  template <typename Vector>
+  void matrix_free_evaluate_agglomerate(dealii::DoFHandler<dim> &dof_handler,
+                                        Vector const &src, Vector &dst) const
+  {
+    dealii::SparsityPattern sparsity_pattern;
+    dealii::SparseMatrix<typename Vector::value_type> system_matrix;
+    dealii::AffineConstraints<double> constraints;
+    this->evaluate_agglomerate(dof_handler, constraints, sparsity_pattern,
+                               system_matrix);
+    system_matrix.vmult(dst, src);
+  }
+
+  // FIXME throw an error to force the user to implement that member function in
+  // the derived class
+  virtual std::vector<double> matrix_free_get_agglomerate_diagonal(
+      dealii::DoFHandler<dim> &dof_handler) const
+  {
+    dealii::AffineConstraints<double> constraints;
+    dealii::SparsityPattern sparsity_pattern;
+    dealii::SparseMatrix<double> system_matrix;
+    this->evaluate_agglomerate(dof_handler, constraints, sparsity_pattern,
+                               system_matrix);
+    size_type const size = dof_handler.n_dofs();
+    std::vector<double> diag_elements(size);
+    for (size_type i = 0; i < size; ++i)
+    {
+      diag_elements[i] = system_matrix.diag_element(i);
+    }
+    return diag_elements;
+  }
+
+  virtual void matrix_free_evaluate_global(
+      dealii::LinearAlgebra::distributed::Vector<double> const &src,
+      dealii::LinearAlgebra::distributed::Vector<double> &dst) const
+  {
+    dealii::DoFHandler<dim> dof_handler;
+    dealii::AffineConstraints<double> constraints;
+    dealii::TrilinosWrappers::SparseMatrix system_matrix;
+    this->evaluate_global(dof_handler, constraints, system_matrix);
+    system_matrix.vmult(dst, src);
+  }
+
+  virtual dealii::LinearAlgebra::distributed::Vector<double>
+  matrix_free_get_diagonal_inverse() const
+  {
+    dealii::DoFHandler<dim> dof_handler;
+    dealii::AffineConstraints<double> constraints;
+    dealii::TrilinosWrappers::SparseMatrix system_matrix;
+    this->evaluate_global(dof_handler, constraints, system_matrix);
+    dealii::IndexSet locally_owned_dofs =
+        system_matrix.locally_owned_domain_indices();
+    dealii::LinearAlgebra::distributed::Vector<double> diagonal(
+        locally_owned_dofs, system_matrix.get_mpi_communicator());
+    for (auto const index : locally_owned_dofs)
+    {
+      diagonal[index] = 1. / system_matrix.diag_element(index);
+    }
+    diagonal.compress(dealii::VectorOperation::insert);
+    return diagonal;
+  }
+};
 
-private:
-  std::shared_ptr<dealii::TrilinosWrappers::SparseMatrix> _sparse_matrix;
-  mutable bool _matrix_initialized = false;
+// Type traits
+template <typename T>
+struct is_matrix_free : std::false_type
```

Comment:
I don't think that this should be in this file.

Reply:
I agree but I am not sure where else to put it.

Reply:
Can we move it to the base class since this should be the default behavior?